### ISSUE TITLE
Highlight FIXME: and TODO: comments

### ISSUE THREAD

**Issue by dominiklohmann** [nvim-treesitter/nvim-treesitter#236]
**Is your feature request related to a problem? Please describe.**

Many colorschemes highlight `FIXME:` and `TODO:` style comments differently. With nvim-treesitter, this is not currently possible.

**Describe the solution you'd like**

Highlight common prefixes like FIXME or TODO within comments.

**Additional context**

Some screenshots w/ [gruvbox colorscheme](https://github.com/gruvbox-community/gruvbox).

With nvim-treesitter: 
<img width="284" alt="image" src="https://user-images.githubusercontent.com/4488655/88665422-e2867080-d0de-11ea-8b7e-04f4c18b02de.png">

Without nvim-treesitter:
<img width="269" alt="image" src="https://user-images.githubusercontent.com/4488655/88665449-eca86f00-d0de-11ea-8c54-dcf587667773.png">

**Comment by vigoux** [comment]
Hi and thanks for reporting this ! What you are aiming for here is what we call language injection, that is we want to treat certain parts of the program (the comments) differently.
We will build a comment (doc and codetags) later and use it together with language injection later. Unfortunately language injection is a bit tougher than expected.

**Comment by dominiklohmann** [comment]
Glad to hear it's on your roadmap already. This is in no way urgent, and I like that you're looking to do this in a generic fashion (which ultimately is what makes treesitter highlighting superior).

**Comment by vigoux** [comment]
Okay, language injection is here now ! Now what we need is a little parser to highlight such code tags, maybe a little code tag parser could be good !

**Comment by dominiklohmann** [comment]
Sounds great. I regularly use the following:
- `FIXME:`, `FIXME(username):` (must be addressed)
- `TODO:`, `TODO(username):` (should be addressed)
- `NYI:` (not yet implemented)

**Comment by vigoux** [comment]
Awesome, it sounds like there was a PEP too specifying that a bit, maybe we can use that as a reference  : https://www.python.org/dev/peps/pep-0350/

There is still some things to do to : for each language we'd have to set up injection, and use the `strip!`operator.

The problem is that the strip operator is not yet implemented and wouldn't word for now, so we need to find a proper way of doing so.

**Comment by theHamsta** [comment]
I planned to do that with https://github.com/nvim-treesitter/nvim-treesitter/pull/590

**Comment by caheredia** [comment]
Likewise the lack of python `TODO` is a pain point for me as well.

**Comment by vigoux** [comment]
So, together with @steelsojka we've been discussing a bit.
There is some upstream issues that might block language injection a bit, and that we need to fix before going forward.

While this work is being done, I'll be working on the `todo` parser, and if anyone wants to contribute, I'll be really grateful.

**Comment by caheredia** [comment]
@vigoux I mostly code in python, but happy to help in any way I can

**Comment by vigoux** [comment]
@steelsojka did a a set of two PRs (neovim/neovim#13207, and https://github.com/nvim-treesitter/nvim-treesitter/pull/627) to be able to inject in a better way.

**Comment by talbergs** [comment]
An absolute MUST feature. See this PR https://github.com/neovim/neovim/pull/13252

I work with files that are `php -> js` and `php -> html -> js` I could determine which of these parser chains to use based on filename.

**Comment by rakin406** [comment]
Yes plssss

**Comment by vigoux** [comment]
Language injection is on by default now, only thing missing is creating the codetag parser, and setting things up for each language.
We are (still) missing some other things like a strip directive, that we will try to hack into here.

**Comment by yashLadha** [comment]
Are there some examples present to achieve this currently with treesitter? @vigoux

**Comment by stsewd** [comment]
I'll try to send a PR tomorrow or next week

![todos](https://user-images.githubusercontent.com/4975310/106377926-7c6cf580-636e-11eb-9cd0-1a6fb48651b4.png)

**Comment by caheredia** [comment]
I see this issue was recently closed, but TODO still isn't working for me in python 

<img width="431" alt="Screen Shot 2021-03-12 at 8 40 00 PM" src="https://user-images.githubusercontent.com/28638529/111019308-3777ae00-8373-11eb-9d47-8ef9f84c2be0.png">
`  
health#nvim_treesitter#check
========================================================================
## Installation
  - OK: `tree-sitter` found  0.19.3(parser generator, used for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found.

## Parser/Features H L F I
  - scala          . . . . 
  - python         ✓ ✓ ✓ ✓ 
  - supercollider  ✓ ✓ ✓ ✓ 
  - go             ✓ ✓ ✓ ✓ 
  - haskell        . . . . 
  - toml           x x x x 
  - glimmer        ✓ . . . 
  - ruby           ✓ ✓ . ✓ 
  - bash           ✓ ✓ ✓ . 
  - vue            ✓ . ✓ . 
  - php            ✓ x ✓ . 
  - java           ✓ ✓ . ✓ 
  - svelte         ✓ . ✓ ✓ 
  - yaml           ✓ ✓ ✓ ✓ 
  - kotlin         ✓ . . . 
  - nix            ✓ ✓ ✓ . 
  - devicetree     ✓ ✓ ✓ ✓ 
  - html           ✓ ✓ ✓ ✓ 
  - dart           x ✓ . ✓ 
  - julia          ✓ ✓ . . 
  - turtle         ✓ ✓ ✓ ✓ 
  - rst            ✓ ✓ . . 
  - json           ✓ ✓ ✓ ✓ 
  - fennel         x ✓ . . 
  - css            x . ✓ ✓ 
  - teal           ✓ ✓ ✓ ✓ 
  - erlang         . . . . 
  - ocaml_interfacex ✓ ✓ . 
  - ql             ✓ x . . 
  - javascript     ✓ x ✓ ✓ 
  - graphql        ✓ . . ✓ 
  - ocaml          x ✓ ✓ . 
  - query          ✓ ✓ ✓ ✓ 
  - elm            . . . . 
  - comment        x . . . 
  - regex          ✓ . . . 
  - c              ✓ ✓ x x 
  - verilog        x x x . 
  - jsdoc          x . . . 
  - ocamllex       x . . . 
  - jsonc          x x x x 
  - swift          . . . . 
  - sparql         x x x x 
  - c_sharp        x . . . 
  - cpp            x x x x 
  - gdscript       x x . . 
  - clojure        x x . . 
  - typescript     x x x x 
  - lua            x x x x 
  - rust           x x x x 
  - tsx            x x x x 
  - ledger         x . x x 
`

**Comment by stsewd** [comment]
@caheredia the syntax is `TODO: something`, you can highlight tokens that don't have `:` if you want with a custom query like

```scm
("text" @text.warning
 (#match? @text.warning "^(TODO|HACK|WARNING)$"))
```

**Comment by stsewd** [comment]
but maybe this something we can add by default? don't know how common is to use just `TODO something` instead of `TODO: something`

**Comment by luc-tielen** [comment]
I would highlight it without the : as well.

**Comment by caheredia** [comment]
> @caheredia the syntax is `TODO: something`, you can highlight tokens that don't have `:` if you want with a custom query like
> 
> ```scheme
@stsewd according to python docs, https://www.python.org/dev/peps/pep-0350/#id25 it looks like Codetags should be followed by a colon. Although in the wild, I've seen more "TODO"s without the colon. Either way thanks for the custom query, I'll add that to my conifgs

**Comment by pabuisson** [comment]
Sorry to reactivate an old issue, but I never got my TODOs and FIXMEs highlighted in my neovim. Currently running 0.6.1, and I'm having the problem in two different languages, ruby and javascript. It's been like this for a while for me, and I was even surprised to find that it was supposed to work.

![Capture d’écran 2022-03-03 à 18 34 21](https://user-images.githubusercontent.com/769918/156619937-7d4ac83e-7368-4191-889c-cd6abcc589a8.png)

Healthcheck seems OK:

```

nvim_treesitter: health#nvim_treesitter#check
========================================================================
## Installation
  - OK: `tree-sitter` found 0.20.4 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v16.13.1 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: Apple clang version 12.0.5 (clang-1205.0.22.11)
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - javascript     ✓ ✓ ✓ ✓ ✓ 
  - ruby           ✓ ✓ ✓ ✓ ✓ 

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

I don't know if color scheme can have an impact (like a non-implemented highlight group?), so I tried to switch and get the same result with nord, onedark, palenight, everforest and basically all decent color schemes I could think of.

Anything I may be missing? At this point I don't know how to investigate any further.

**Comment by clason** [comment]
You need to install the `comment` parser to get highlighting inside comments.

**Comment by pabuisson** [comment]
Damn, I didn't understand this 🤦‍♂️ Maybe it'll also help other people someday.
Sorry for the noise and thanks a lot for the help!

**Comment by diegoulloao** [comment]
> 

How I do that?

**Comment by diegoulloao** [comment]
> You need to install the `comment` parser to get highlighting inside comments.

How I do that?

**Comment by diegoulloao** [comment]
Now I realized, I just added `comment` inside the `ensure_installed` list in my treesitter configuration.
Anyway, I think there are better solutions for this purpose: https://github.com/folke/todo-comments.nvim

**Comment by joellidin** [comment]
I am sorry to yet again bring this up. But for python it still does not highlight the todo comments for me. 
![image](https://user-images.githubusercontent.com/19657866/198013599-1a345c0d-d8ae-46c3-afc0-4b33f08160cf.png)
 The healtcheck seems fine:
```
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - WARNING: `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
  - OK: `node` found v14.20.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: cc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
  - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - go             ✓ ✓ ✓ ✓ ✓
  - bash           ✓ ✓ ✓ . ✓
  - cpp            ✓ ✓ ✓ ✓ ✓
  - cuda           ✓ ✓ ✓ ✓ ✓
  - regex          ✓ . . . .
  - graphql        ✓ . . ✓ ✓
  - markdown       ✓ . ✓ . ✓
  - make           ✓ . ✓ . ✓
  - python         ✓ ✓ ✓ ✓ ✓
  - rust           ✓ ✓ ✓ ✓ ✓
  - sql            ✓ . . . ✓
  - typescript     ✓ ✓ ✓ ✓ ✓
  - json           ✓ ✓ ✓ ✓ .
  - comment        ✓ . . . .
  - lua            ✓ ✓ ✓ ✓ ✓
  - dockerfile     ✓ . . . ✓
  - c              ✓ ✓ ✓ ✓ ✓
  - scss           ✓ . . ✓ .
  - vim            ✓ ✓ ✓ . ✓
  - toml           ✓ ✓ ✓ ✓ ✓
  - yaml           ✓ ✓ ✓ ✓ ✓
  - javascript     ✓ ✓ ✓ ✓ ✓
  - bibtex         ✓ . ✓ ✓ .
  - css            ✓ . ✓ ✓ ✓
  - latex          ✓ . ✓ . ✓
  - help           ✓ . . . .
```

**Comment by stsewd** [comment]
@joellidin I see that the `:` are highlighted, so this is probably your colorscheme missing highlighting the `@text.*` captures, see https://github.com/nvim-treesitter/nvim-treesitter/issues/2293#issuecomment-1279974776

**Comment by joellidin** [comment]
> @joellidin I see that the `:` are highlighted, so this is probably your colorscheme missing highlighting the `@text.*` captures, see [#2293 (comment)](https://github.com/nvim-treesitter/nvim-treesitter/issues/2293#issuecomment-1279974776)

@stsewd Thank you for the response. I looked and it seems that my colorscheme should have it. See picture below when I checked `:hi`. Worth noting maybe is that the todo comments get highlighted when I disable TS hightlighting. Is it anything obvious I am missing maybe?
![image](https://user-images.githubusercontent.com/19657866/198275435-acda05f3-9cb3-4d3c-8d65-1fc099485989.png)
